lib/aws/signature.ex

```diff
@@ -14,229 +10,19 @@ defmodule AWS.Signature do
-      |> add_date_header(long_date)
-      |> add_content_hash(body)
-      |> add_security_token(client)
-
-    canonical_request = canonical_request(method, url, headers, body)
-    hashed_canonical_request = Util.sha256_hexdigest(canonical_request)
-    credential_scope = credential_scope(short_date, region, client.service)
-
-    signing_key = signing_key(client, short_date)
-
-    string_to_sign = string_to_sign(long_date, credential_scope, hashed_canonical_request)
-
-    signature = Util.hmac_sha256_hexdigest(signing_key, string_to_sign)
-    signed_headers = signed_headers(headers)
-
-    authorization =
-      authorization(client.access_key_id, credential_scope, signed_headers, signature)
-
-    add_authorization_header(headers, authorization)
-  end
-
-  @doc """
-  Generate headers with an AWS signature version 4 for the specified request using the specified time that can be transformed into a query string.
-  """
-  def sign_v4_query(client, now, method, url, headers, body) do
-    long_date = NaiveDateTime.to_iso8601(now, :basic) <> "Z"
-    short_date = Date.to_iso8601(now, :basic)
-    headers = add_date_header(headers, long_date)
-    canonical_request = canonical_request(method, url, headers, body)
-    hashed_canonical_request = Util.sha256_hexdigest(canonical_request)
-    region = client.region || @default_region_for_global_services
-
-    credential_scope = credential_scope(short_date, region, client.service)
-
-    signing_key = signing_key(client, short_date)
-
-    string_to_sign = string_to_sign(long_date, credential_scope, hashed_canonical_request)
-
-    signature = Util.hmac_sha256_hexdigest(signing_key, string_to_sign)
-    signed_headers = signed_headers(headers)
-
-    credential =
-      Enum.join(
-        [client.access_key_id, short_date, region, client.service, "aws4_request"],
-        "/"
-      )
-
-    result = [
-      {"X-Amz-Algorithm", "AWS4-HMAC-SHA256"},
-      {"X-Amz-Credential", credential},
-      {"X-Amz-Date", long_date},
-      {"X-Amz-SignedHeaders", signed_headers},
-      {"X-Amz-Signature", signature}
-    ]
-
-    if expiry = :proplists.get_value("X-Amz-Expires", headers, nil) do
-      [{"X-Amz-Expires", expiry} | result]
-    else
-      result
-    end
-  end
-
-  defp add_date_header(headers, date) do
-    [{"X-Amz-Date", date} | headers]
-  end
-
-  # Add an X-Amz-Content-SHA256 header which is the hash of the payload.
-  # This header is required for S3 when using the v4 signature. Adding it
-  # in requests for all services does not cause any issues.
-  defp add_content_hash(headers, body) do
-    [{"X-Amz-Content-SHA256", AWS.Util.sha256_hexdigest(body)} | headers]
-  end
-
-  # Add an `X-Amz-Security-Token` if credentials configurations are configured for it
-  defp add_security_token(headers, %AWS.Client{session_token: nil}), do: headers
-
-  defp add_security_token(headers, %AWS.Client{session_token: session_token}),
-    do: [{"X-Amz-Security-Token", session_token} | headers]
```

Comment:
`:aws_signature` doesn't add this header, so we need to add it conditionally before the signature.

Reasoning: this is pretty much an arbitrary, optional header and it's not relevant to the signing process, so it should be added manually beforehand as any other header :)

Reply:
@jonatanklosko good catch! I didn't notice it.